### PR TITLE
Add a function that conditions reference density in DDEC6

### PR DESCRIPTION
**Pull request by mscho527**
_Related Issue: #885
Depends upon: #916_

This PR adds a functionality to apply conditioning for the proatomic densities and thus implements the third step of DDEC6 algorithm. Tests ensure that the integrated charges of conditioned densities match expected values and that tau values (equation 66) match values reported by `chargemol`.

### CODE REVIEW

File: cclib/method/volume.py [hunk 1]
```diff
@@ -178,23 +178,28 @@ def writeasvtk(self, filename):
         )
         v.tofile(filename)
 
-    def integrate(self):
+    def integrate(self, weights=None):
+        weights = numpy.ones_like(self.data) if weights is None else weights
```
Comment (shivupa) [inline cclib/method/volume.py:182]: can we assert that weights is the right size

Comment (mscho527) [reply cclib/method/volume.py:182]: Yes -- I'll add it in. Thanks!

File: cclib/method/ddec.py [hunk 2]
```diff
@@ -300,6 +327,324 @@ def calculate_refcharges(self):
...
             # (Equation 50 and 58 in doi: 10.1039/c6ra04656h)
-            refcharges[atomi] = (stockholdercharges[atomi] / 3.0) + (localizedcharges[atomi] * 2.0 / 3.0)
-        
+            refcharges[atomi] = (stockholdercharges[atomi] / 3.0) + (
+                localizedcharges[atomi] * 2.0 / 3.0
+            )
+
         return refcharges, localizedcharges, stockholdercharges
+
+    def condition_densities(self):
```
Comment (berquist) [inline cclib/method/ddec.py:336]: Docstring?

File: cclib/method/ddec.py [hunk 3]
```diff
@@ -300,6 +327,324 @@ def calculate_refcharges(self):
...
+            for z in range(zshape)
+        )
+
+        self._rho_ref = numpy.zeros((xshape, yshape, zshape))
+
+        for atomi, xindex, yindex, zindex in indices:
+            # rho_ref -- Equation 41 in doi: 10.1039/c6ra04656h
+            self._rho_ref[xindex][yindex][zindex] = (
+                self._rho_ref[xindex][yindex][zindex]
+                + self.proatom_density[atomi][self.closest_r_index[atomi][xindex][yindex][zindex]]
```
Comment (berquist) [inline cclib/method/ddec.py:354]: Can you do a += here?

File: cclib/method/ddec.py [hunk 4]
```diff
@@ -300,6 +327,324 @@ def calculate_refcharges(self):
...
+            )
+
+        self._candidates_bigPsi = []
```
Comment (berquist) [inline cclib/method/ddec.py:357]: All the symbols in the SI are actually big and little phi (excluding rho).

File: cclib/method/ddec.py [hunk 5]
```diff
@@ -300,6 +327,324 @@ def calculate_refcharges(self):
...
+        grid = copy.deepcopy(self.chgdensity)
+        grid.data = numpy.zeros_like(grid.data)
+
+        xshape, yshape, zshape = self.chgdensity.data.shape
+        indices = ((x, y, z) for x in range(xshape) for y in range(yshape) for z in range(zshape))
+
+        for density, atomi in zip(radial_density_list, atom_list):
+            for x, y, z in indices:
+                grid.data[x][y][z] = (
+                    grid.data[x][y][z] + density[self.closest_r_index[atomi][x][y][z]]
```
Comment (berquist) [inline cclib/method/ddec.py:647]: += here

File: cclib/method/ddec.py [hunk 6]
```diff
@@ -300,6 +327,324 @@ def calculate_refcharges(self):
...
+            else:
+                spherical_avg = numpy.sum(grid_filter * weights) / num_grid_filter
+                ya[radiusi] = proatom_density[atomi][radiusi] * spherical_avg
+
+        ya = numpy.maximum.accumulate(ya[::-1])[::-1]
+        ya = numpy.minimum.accumulate(ya)
+
+        # Normalize y_a (see module_DDEC6_valence_iterator.f08::284)
+        nelec = self._integrate_from_radial([ya], [atomi])
+        ya = ya * (self.data.atomnos[atomi] - self.refcharges[-1][atomi]) / nelec
```
Comment (berquist) [inline cclib/method/ddec.py:631]: *= here

File: cclib/method/ddec.py [hunk 7]
```diff
@@ -300,6 +327,324 @@ def calculate_refcharges(self):
...
+        for radiusi in range(len(ya)):
+            grid_filter = self.closest_r_index[atomi] == radiusi
+            num_grid_filter = numpy.count_nonzero(grid_filter)
+            if num_grid_filter < 1:
+                ya[radiusi] = 0.0
+            else:
+                spherical_avg = numpy.sum(grid_filter * weights) / num_grid_filter
+                ya[radiusi] = proatom_density[atomi][radiusi] * spherical_avg
+
+        ya = numpy.maximum.accumulate(ya[::-1])[::-1]
```
Comment (berquist) [inline cclib/method/ddec.py:626]: Can you explain why you need to reverse, accumulate, then reverse again?

File: cclib/method/ddec.py [hunk 8]
```diff
@@ -300,6 +327,324 @@ def calculate_refcharges(self):
...
+        return psiAI, cond_density
+
+    def _ya(self, proatom_density, atomi):
+        # Function that calculates Y_a^avg
+        # See Eq. 40-41 in doi: 10.1039/c6ra04656h
+        rho_ref = self._rho_ref
+        # Y_a^avg -- Equation 40 in doi: 10.1039/c6ra04656h
+        ya = numpy.zeros_like(proatom_density[atomi], dtype=float)
+        weights = self.chgdensity.data / rho_ref
+        summed_elec = 0
```
Comment (berquist) [inline cclib/method/ddec.py:616]: Unused variable?

File: cclib/method/ddec.py [hunk 9]
```diff
@@ -300,6 +327,324 @@ def calculate_refcharges(self):
...
+                        self.tau[atomi][radiusi] = numpy.divide(
+                            leftaverage,
+                            rightaverage,
+                            out=numpy.zeros_like(leftaverage),
+                            where=rightaverage != 0.0,
+                        )
+            # Make tau monotonic decreasing
+            self.tau[atomi] = numpy.maximum.accumulate(self.tau[atomi][::-1])[::-1]
+
+    def _psiai(self, ya, bigpsiAI, atomi):
```
Comment (berquist) [inline cclib/method/ddec.py:593]: Docstring for where this comes from (second box in S1?)

File: cclib/method/ddec.py [hunk 10]
```diff
@@ -300,6 +327,324 @@ def calculate_refcharges(self):
...
+            psiAI[atomi] = (
+                self._integrate_from_radial([self._cond_density[atomi]], [atomi])
+                - self.data.atomnos[atomi]
+                + self.refcharges[-1][atomi]
+            )
+
+            self._candidates_bigPsi.append([bigpsiAI[atomi]])
+            self._candidates_psi.append([psiAI[atomi]])
+
+            fitpsi = True  # when convergence is reached, this is modified to False.
```
Comment (shivupa) [inline cclib/method/ddec.py:385]: Sorry to nitpick but as I was going through and checking the equations I realized these are phi not psi. Can we change that to stay consistent with the paper?

Comment (shivupa) [reply cclib/method/ddec.py:385]: (This applies throughout this function)

File: cclib/method/ddec.py (hunk 9 above)
Comment (berquist) [reply cclib/method/ddec.py:593]: Oh, it's S100, better to say that.

File: cclib/method/ddec.py [hunk 11]
```diff
@@ -300,6 +327,324 @@ def calculate_refcharges(self):
...
+                )
+                if numpy.count_nonzero(self._candidates_psi[atomi] < 0) > 0:
+                    lower_ind = numpy.where(
+                        self._candidates_psi[atomi]
+                        == self._candidates_psi[atomi][self._candidates_psi[atomi] < 0].max()
+                    )[0][0]
+                    lowerbigPsi = self._candidates_bigPsi[atomi][lower_ind]
+                    lowerpsi = self._candidates_psi[atomi][lower_ind]
+                else:  # assign some large negative number
+                    lowerbigPsi = -9999999.0
```
Comment (berquist) [inline cclib/method/ddec.py:422]: Can you use `numpy.NINF`?

File: cclib/method/ddec.py [hunk 12]
```diff
@@ -300,6 +327,324 @@ def calculate_refcharges(self):
...
+                    lowerpsi = -9999999.0
+                if numpy.count_nonzero(self._candidates_psi[atomi] > 0) > 0:
+                    upper_ind = numpy.where(
+                        self._candidates_psi[atomi]
+                        == self._candidates_psi[atomi][self._candidates_psi[atomi] > 0].min()
+                    )[0][0]
+                    upperbigPsi = self._candidates_bigPsi[atomi][upper_ind]
+                    upperpsi = self._candidates_psi[atomi][upper_ind]
+                else:  # assign some large positive number
+                    upperbigPsi = 9999999.0
```
Comment (shivupa) [inline cclib/method/ddec.py:432]: can we use something like np.inf here and above

Comment (berquist) [inline cclib/method/ddec.py:432]: This would be `numpy.PINF`.

File: cclib/method/ddec.py (hunk 10 above)
Comment (mscho527) [reply cclib/method/ddec.py]: I've changed the variables into phi. Thanks for noting this!

File: cclib/method/ddec.py (hunk 12 above)
Comment (mscho527) [reply cclib/method/ddec.py]: I've used np.PINF and np.NINF which were suggested by Eric in a comment below this. Thanks!

File: cclib/method/ddec.py (hunk 2 above)
Comment (mscho527) [reply cclib/method/ddec.py:339]: Looks like I've added the docstring only to the later (dependent) PRs of this. If added here as well. Thanks!

File: cclib/method/ddec.py (hunk 3 above)
Comment (mscho527) [reply cclib/method/ddec.py]: Yes! I've tried to change any other similar cases as well. Thanks!

File: cclib/method/ddec.py (hunk 5 above)
Comment (mscho527) [reply cclib/method/ddec.py:652]: Yes! I've tried to change any other similar cases as well. Thanks!

File: cclib/method/ddec.py (hunk 6 above)
Comment (mscho527) [reply cclib/method/ddec.py]: Yes! I've tried to change any other similar cases as well. Thanks!

File: cclib/method/ddec.py (hunk 4 above)
Comment (mscho527) [reply cclib/method/ddec.py]: I've changed this. Thanks!

File: cclib/method/ddec.py (hunk 8 above)
Comment (mscho527) [reply cclib/method/ddec.py]: I must have forgot to remove this. Thanks!

File: cclib/method/ddec.py (hunk 7 above)
Comment (mscho527) [reply cclib/method/ddec.py:631]: This was in the `module_reshaping_functions.f08:77-79`. I've added a comment about this. Thanks.

File: cclib/method/ddec.py (hunk 9 above)
Comment (mscho527) [reply cclib/method/ddec.py]: I've added this. Thanks!

File: cclib/method/ddec.py (hunk 11 above)
Comment (mscho527) [reply cclib/method/ddec.py]: Sure. This looks much neater! Thanks.

File: cclib/method/ddec.py (hunk 12 above)
Comment (mscho527) [reply cclib/method/ddec.py]: Sure. This looks much neater! Thanks.

**Pull request merged**